firedrake/slate/slac/compiler.py

```diff
@@ -41,18 +45,147 @@
+cell_orientations_sym = ast.Symbol("cell_orientations")
+cell_facet_sym = ast.Symbol("cell_facets")
+it_sym = ast.Symbol("i0")
+mesh_layer_sym = ast.Symbol("layer")
+
+# Supported integral types
 supported_integral_types = [
     "cell",
     "interior_facet",
     "exterior_facet",
-    "interior_facet_horiz",
+    # The "interior_facet_horiz" measure is separated into two parts:
+    # "top" and "bottom"
+    "interior_facet_horiz_top",
+    "interior_facet_horiz_bottom",
     "interior_facet_vert",
     "exterior_facet_top",
     "exterior_facet_bottom",
     "exterior_facet_vert"
 ]
 
 
+class LocalKernelBuilder(KernelBuilderBase):
+    """The primary helper class for constructing cell-local linear
+    algebra kernels from Slate expressions.
+
+    This class provides access to all temporaries and subkernels associated
+    with a Slate expression. If the Slate expression contains nodes that
+    require operations on already assembled data (such as the action of a
+    slate tensor on a `ufl.Coefficient`), this class provides access to the
+    expression which needs special handling.
+
+    Instructions for assembling the full kernel AST of a Slate expression is
+    provided by the method `construct_ast`.
+    """
+    def __init__(self, expression, tsfc_parameters=None):
+        """Constructor for the LocalKernelBuilder class.
+
+        :arg expression: a :class:`TensorBase` object.
+        :arg tsfc_parameters: an optional `dict` of parameters to provide to
+                              TSFC when constructing subkernels associated
+                              with the expression.
+        """
+        super(LocalKernelBuilder, self).__init__(expression=expression,
+                                                 tsfc_parameters=tsfc_parameters)
+        transformer = Transformer()
+        include_dirs = []
+        templated_subkernels = []
+        assembly_calls = OrderedDict([(it, []) for it in supported_integral_types])
+        coords = None
+        oriented = False
+        for cxt_kernel in self.context_kernels:
+            local_coefficients = cxt_kernel.coefficients
+            it_type = cxt_kernel.original_integral_type
+            exp = cxt_kernel.tensor
+
+            if it_type not in supported_integral_types:
+                raise ValueError("Integral type '%s' not recognized" % it_type)
+
+            # Explicit checking of coordinates
+            coordinates = cxt_kernel.tensor.ufl_domain().coordinates
+            if coords is not None:
+                assert coordinates == coords, "Mismatching coordinates!"
+            else:
+                coords = coordinates
+
+            for split_kernel in cxt_kernel.tsfc_kernels:
+                indices = split_kernel.indices
+                kinfo = split_kernel.kinfo
+
+                # TSFC No-op kernels do not contain a kernel body. If this is
+                # the case, we don't create a function call.
+                if kinfo.kernel:
```

Comment:
When is it the case that `compile_form` returns an iterable of kernels from tsfc that contains an empty kernel?

Reply:
This only really occurs when using the `HybridizationPC`. Suppose `a` is the mixed bilinear form describing the mixed poisson operator. When generating the reconstruction calls for the flux and scalar unknowns, the form is split to generate the relevant local `Tensor` objects. Calling `split_form(a)` will zero-pad the scalar-mass bit (since it doesn't exist in the Poisson case) and when passing that through TSFC you get a no-op. The kernel bodies for these are empty.

I implemented it this way because I wanted to keep the PC somewhat general. Whether you're solving a mixed Poisson/Helmholtz, or one reverses the typical ordering of the function spaces (HDiv * L^2 or L^2 * HDiv), the reconstruction expressions work for all cases, except some terms may be "zero."

Ideally, the best way to handle this would be to create some sort of `Visitor` that removes all the `noop` (zero-valued tensors) nodes in an expression and then reconstruct the expression with those nodes removed. However, I haven't implemented such a thing and I'll need some time to figure that out. I figured this is good enough for now.

Reply:
I don't understand, you make `cxt_kernel.tsfc_kernels` like this:

`cxt_kernel.tsfc_kernels = compile_form(form)`

if `form` evaluates to a no-op, then `compile_form` returns an empty tuple, so the iteration `for split_kernel in cxt_kernel.tsfc_kernels` will never execute the body.

e.g.
```python
mesh = UnitSquareMesh(1, 1)
V = FunctionSpace(mesh, BrokenElement(FiniteElement("RT", triangle, 1)))

Q = FunctionSpace(mesh, "DG", 0)

W = V*Q

u, p = TrialFunctions(W)

v, q = TestFunctions(W)

a = (dot(u, v) + div(u)*q + div(v)*p)*dx

from firedrake.formmanipulation import split_form

splits = dict(split_form(a))
compile_form(splits[(1, 1)], "foo") => ()
```

Or am I completely missing something?

Reply:
No you're right, I am confusing a time when I was messing around with TSFC split kernels using methods imported directly from TSFC, not `tsfc_interface.py`. I'll remove the special casing here.

However, the "zero-tensor" thing is still an issue. While no assembly calls are made, a `Tensor` temporary with the right shape but initialised as zero is still created and used in the generated code.

Reply:
That suggests that you need to prune earlier when you compile the form and notice that it's zero.

Reply:
Yes, but once I have an expression, and I know a particular operand is zero, I'm not sure how to remove the node and reconstruct the expression. If `O` is a zero tensor, I want to transform: `A * O + B.T * D.inv * B - O` into `B.T * D.inv * B`. I need to have a think about how to do this in code.

Reply:
OK, I see.  I think this is the wrong time to do that.  Here's an idea (not for this PR).  You split the expression compilation into multiple passes.  In the first pass, you visit the terminals of the DAG and  compile the form using tsfc, you then reconstruct a new type of terminal with this compiled form.  At this point when you reconstruct the DAG, you can put any zero simplification in.

Now you have a DAG where there are no forms needed to compile and the next pass just needs to glue to kernel calls together.

This is morally quite similar to how tsfc translates from UFL.  First, replace finite element basis function evaluation with their finat representation, then translate the gem to code.